ignite/metrics/hsic.py

```diff
@@ -0,0 +1,169 @@
+
+        .. testoutput::
+
+            0.0922664999961853
+
+    .. versionchanged:: 0.5.1
+        ``skip_unrolling`` argument is added.
+    """
+
+    def __init__(
+        self,
+        sigma_x: float = -1,
+        sigma_y: float = -1,
+        ignore_invalid_batch: bool = True,
+        output_transform: Callable = lambda x: x,
+        device: Union[str, torch.device] = torch.device("cpu"),
+        skip_unrolling: bool = False,
+    ):
+        super().__init__(output_transform, device, skip_unrolling=skip_unrolling)
+
+        self.sigma_x = sigma_x
+        self.sigma_y = sigma_y
+        self.ignore_invalid_batch = ignore_invalid_batch
+
+    _state_dict_all_req_keys = ("_sum_of_trace", "_sum_of_second_term", "_sum_of_third_term", "_num_batches")
+
+    @reinit__is_reduced
+    def reset(self) -> None:
+        self._sum_of_trace = torch.tensor(0.0, device=self._device)
+        self._sum_of_second_term = torch.tensor(0.0, device=self._device)
+        self._sum_of_third_term = torch.tensor(0.0, device=self._device)
+        self._num_batches = 0
+
+    @reinit__is_reduced
+    def update(self, output: Sequence[Tensor]) -> None:
+        X = output[0].detach().flatten(start_dim=1)
+        Y = output[1].detach().flatten(start_dim=1)
+        b = X.shape[0]
+
+        if b <= 3:
+            if self.ignore_invalid_batch:
+                return
+            else:
+                raise ValueError(f"A batch must contain more than four samples, got only {b} samples.")
+
+        mask = 1.0 - torch.eye(b, device=X.device)
+
+        xx = X @ X.T
+        rx = xx.diag().unsqueeze(0).expand_as(xx)
+        dxx = rx.T + rx - xx * 2
+        if self.sigma_x < 0:
+            vx = torch.quantile(dxx, 0.5)
+        else:
+            vx = self.sigma_x**2
+        K = torch.exp(-0.5 * dxx / vx) * mask
+
+        yy = Y @ Y.T
+        ry = yy.diag().unsqueeze(0).expand_as(yy)
+        dyy = ry.T + ry - yy * 2
+        if self.sigma_y < 0:
+            vy = torch.quantile(dyy, 0.5)
+        else:
+            vy = self.sigma_y**2
+        L = torch.exp(-0.5 * dyy / vy) * mask
+
+        KL = K @ L
+        trace = KL.trace() / (b * (b - 3))
+        second_term = K.sum() * L.sum() / (b * (b - 1) * (b - 2) * (b - 3))
+        third_term = KL.sum() / (b * (b - 2) * (b - 3))
+
+        self._sum_of_trace += trace.to(self._device)
+        self._sum_of_second_term += second_term.to(self._device)
+        self._sum_of_third_term += third_term.to(self._device)
+
+        self._num_batches += 1
+
+    @sync_all_reduce("_sum_of_trace", "_sum_of_second_term", "_sum_of_third_term", "_num_batches")
+    def compute(self) -> float:
+        if self._num_batches == 0:
+            raise NotComputableError("HSIC must have at least one batch before it can be computed.")
+
+        hsic = self._sum_of_trace + self._sum_of_second_term - self._sum_of_third_term * 2.0
+        hsic = hsic.clamp(min=0.0)  # HSIC must not be negative
```

Comment:
I wonder whether in the paper they say somewhere to clip the value ?

Reply:
Thanks for your review!

Although they do not use clipping, I observed the computation sometimes produced slight negative values because of numerical errors.
So, I added clipping because the HSIC is the norm of the covariance in definition (Eq. (3) of Song, et al. (2012)), and it must not be negative.